db/db_impl/db_impl.cc

```diff
@@ -522,7 +523,22 @@ Status DBImpl::CloseHelper() {
   int flushes_unscheduled = env_->UnSchedule(this, Env::Priority::HIGH);
   Status ret = Status::OK();
   mutex_.Lock();
+
   bg_bottom_compaction_scheduled_ -= bottom_compactions_unscheduled;
+  if (bg_flush_scheduled_low_) {
```

Comment:
I also found the same issue and attempt to fix in #8007. Here I have a doubt. It looks you have an assumption that all bg_flush_scheduled_low_ will be successfully unscheduled, but there could be flushes not in unscheduled queue yet. Otherwise, the original logic could become something as below if following this assumption 
```
  bg_compaction_scheduled_ = 0;
  bg_flush_scheduled_ = 0;
```
Actually I met store hanging in DBTest.DynamicMemtableOptions unit test, and I can repro the issue with this test case on Windows. With your fix, I have the following assert error after running many times
```
Note: Google Test filter = DBTest.DynamicMemtableOptions
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from DBTest
[ RUN      ] DBTest.DynamicMemtableOptions
Assertion failed: bg_flush_scheduled_, file e:\repo\rocksdb\db\db_impl\db_impl_compaction_flush.cc, line 2434
```
If you check that line, it means `bg_flush_scheduled_` is already 0 but there is still background flush running.

Reply:
@sherriiiliu  Thanks for pointing it out. I will look into it and update the fix.

Reply:
@sherriiiliu  Can you share the Assertion failure. Also, bg_flush_scheduled_low_ represents jobs that are in the queue but not running yet.  I didn't fully understand your point here:                                       
 ```bg_compaction_scheduled_ = 0; // all compactions are unscheduled in LOW pool
  bg_flush_scheduled_ = 0; // all flushes are unscheduled in HIGH pool```
Jobs which are unscheduled. they are not in the queue. They are represented by ```unscheduled_flushes_```

Reply:
One more thing you mentioned "It looks you have an assumption that all bg_flush_scheduled_low_ will be successfully unscheduled, but there could be flushes not in unscheduled queue yet." But then don't we need to consider those because they are not running/haven't be scheduled yet.``` unscheduled_flushes_``` would represent those jobs.

Reply:
The assertion failure is in `DBImpl::BackgroundCallFlush(Env::Priority thread_pri)`:
```
void DBImpl::BackgroundCallFlush(Env::Priority thread_pri) {
  bool made_progress = false;
  JobContext job_context(next_job_id_.fetch_add(1), true);

  TEST_SYNC_POINT("DBImpl::BackgroundCallFlush:start");

  LogBuffer log_buffer(InfoLogLevel::INFO_LEVEL,
                       immutable_db_options_.info_log.get());
  {
    InstrumentedMutexLock l(&mutex_);
    assert(bg_flush_scheduled_); // fails here, bg_flush_scheduled_=0
    ...
```
I am not sure if you can repro on Linux or MaxOS, but it happens on my Win10.
My point is, the value returned by `env_->UnSchedule(this, Env::Priority::LOW)` is not always equal to the scheduled jobs in the corresponding threadpool by my observation. I think that's the reason why the original code logic is to decrease `bg_compaction_scheduled_ ` or `bg_flush_scheduled_` by the actual number that were successfully unscheduled, and then check if `bg_compaction_scheduled_ ` or `bg_flush_scheduled_` is greater than 0 to see if we need to wait any leftover jobs to finish in the while loop. Here you assume if there is any flush job in LOW pool, `bg_flush_scheduled_low_` will become 0 after unschedule is triggered, which looks contradict the above logic.

Reply:
This is a good point. Only the portion of ```bg_flush_scheduled_low_``` that's still in the thread pool queue will be returned by ```env->UnSchedule```. Therefore, adjust ```compactions_unscheduled``` and ```flushes_unscheduled``` using only ```bg_flush_scheduled_low_```.

How about considering ```bg_compaction_scheduled_ + bg_flush_scheduled_``` as one metric? Then you just have to subtract ```compactions_unscheduled + flushes_unscheduled``` and not have to track ```bg_flush_scheduled_low_```.

Reply:
Exactly what I think. In my fix #8007 , I only set `num_low_scheduled` to track the sum of `bg_compaction_scheduled_ + bg_flush_scheduled_`